### PR TITLE
Add client fingerprint to open and close assessment events

### PR DESCRIPTION
**Pull request by mylesw27**
This PR seeks to add the client fingerprint ID to the `assessment_state_logs` table when an assessment is opened or closed. This data will then be used to show the client fingerprint data in the event logs of assessment instance.

### CODE REVIEW

File: apps/prairielearn/src/middlewares/clientFingerprint.ts [hunk 1]
```diff
@@ -52,3 +52,31 @@ export default asyncHandler(async (req, res, next) => {
...
+    params,
+    IdSchema,
+  );
+
+  if (!client_fingerprint_id) {
+    client_fingerprint_id = await sqldb.queryRow(sql.insert_client_fingerprint, params, IdSchema);
+  }
+
+  return client_fingerprint_id;
+}
```
Comment (mylesw27) [inline apps/prairielearn/src/middlewares/clientFingerprint.ts:82]: Adding an additional function here was necessary as the assessment instance ID has not been created yet when accessing from `studentAssessmentExam` or `studentAssessmentHomework`. For that reason, we cannot use the default function from the `clientFingerprints` middleware.

File: apps/prairielearn/src/pages/studentAssessmentExam/studentAssessmentExam.js [hunk 2]
```diff
@@ -127,6 +128,7 @@ router.post(
       // students to create and start a new assessment instance.
       if (!checkPasswordOrRedirect(req, res)) return;
 
+      const client_fingerprint_id = await getClientFingerprintId(req, res);
```
Comment (mylesw27) [inline apps/prairielearn/src/pages/studentAssessmentExam/studentAssessmentExam.js:131]: We did not want to use `clientFingerprints` middleware on the entire route. We only want to access the client fingerprint if, and only if, there is a post request to open the assessment. This prevents unwarranted fingerprints from viewing the assessment page before it has started.

File: apps/prairielearn/src/middlewares/clientFingerprint.ts [hunk 3]
```diff
@@ -52,3 +52,31 @@ export default asyncHandler(async (req, res, next) => {
   res.locals.client_fingerprint_id = client_fingerprint_id;
   next();
 });
+
+export async function getClientFingerprintId(req: any, res: any) {
```
Comment (nwalters512) [inline apps/prairielearn/src/middlewares/clientFingerprint.ts:56]: This function does largely the same thing as the above middleware. I'd expect that we could reduce duplication here, possibly by using `getClientFingerrpintId` in the above middleware.

File: apps/prairielearn/src/cron/autoFinishExams.ts [hunk 4]
```diff
@@ -40,6 +40,7 @@ export function run(callback: (err?: Error | null) => void): void {
           requireOpen,
           examItem.close_assessment,
           overrideGradeRate,
+          null, //client_fingerprint_id
```
Comment (nwalters512) [inline apps/prairielearn/src/cron/autoFinishExams.ts:43]: Nit: the pattern with the other arguments here was to declare them in a local variable that we can add a comment to, consider doing the same here (though this is by no means a blocker).

File: apps/prairielearn/src/middlewares/clientFingerprint.ts (hunk 3 above)
Comment (nwalters512) [inline apps/prairielearn/src/middlewares/clientFingerprint.ts:56]: There are `Request` and `Response` types in the `express` package that you can use for the arguments here.

Comment (mylesw27) [reply apps/prairielearn/src/middlewares/clientFingerprint.ts]: That makes sense. Added use of `getClientFingerprintId` to the middleware to reduce code duplication in commit [916c5a5](https://github.com/PrairieLearn/PrairieLearn/pull/9087/commits/916c5a53cfa2b70e94b9b2dc43be9bdc3529fd6f).

File: apps/prairielearn/src/cron/autoFinishExams.ts (hunk 4 above)
Comment (mylesw27) [reply apps/prairielearn/src/cron/autoFinishExams.ts]: Declared a local variable `clientFingerprintId` with a null value and added a comment for it in commit [916c5a5](https://github.com/PrairieLearn/PrairieLearn/pull/9087/commits/916c5a53cfa2b70e94b9b2dc43be9bdc3529fd6f).

File: apps/prairielearn/src/middlewares/clientFingerprint.ts (hunk 3 above)
Comment (mylesw27) [reply apps/prairielearn/src/middlewares/clientFingerprint.ts]: Added use of `Request` and `Response` types in commit [916c5a5](https://github.com/PrairieLearn/PrairieLearn/pull/9087/commits/916c5a53cfa2b70e94b9b2dc43be9bdc3529fd6f).

File: apps/prairielearn/src/middlewares/clientFingerprint.ts [hunk 5]
```diff
@@ -18,11 +36,6 @@ export default asyncHandler(async (req, res, next) => {
 
   const params = {
     ip_address: req.ip,
-    // We are passing the authn user id here. However, we are checking
-    // in the SQL query 'update_assessment_instance_fingerprint'
-    // that the authn user is the owner of the assessment. This will keep us
-    // from inadvertently recording a fingerprint change
-    // for an instructor viewing the assessment instance.
```
Comment (nwalters512) [inline apps/prairielearn/src/middlewares/clientFingerprint.ts:25]: This is a useful comment; could we keep it around somewhere?

File: apps/prairielearn/src/pages/instructorAssessmentInstances/instructorAssessmentInstances.js [hunk 6]
```diff
@@ -106,6 +106,7 @@ router.post(
         requireOpen,
         close,
         overrideGradeRate,
+        null, //client_fingerprint_id
```
Comment (nwalters512) [inline apps/prairielearn/src/pages/instructorAssessmentInstances/instructorAssessmentInstances.js:109]: Formatting nit (we conventionally put leading spaces on comments):

```suggestion
        null, // client_fingerprint_id
```

File: apps/prairielearn/src/middlewares/clientFingerprint.ts (hunk 5 above)
Comment (mylesw27) [reply apps/prairielearn/src/middlewares/clientFingerprint.ts]: Yes, this was lost in the process of refactoring functions to reduce duplication. I have now restored this comment to the same place in 52d0e1a.

File: apps/prairielearn/src/pages/instructorAssessmentInstances/instructorAssessmentInstances.js (hunk 6 above)
Comment (mylesw27) [reply apps/prairielearn/src/pages/instructorAssessmentInstances/instructorAssessmentInstances.js]: Added this leading space in commit 52d0e1a.

**Review by nwalters512 — APPROVED** [review]

**Pull request merged**